More tests, add `filename` to Torrent, fast Database#include? method

winterthediplomat opened the pull request:
I'm sorry I did not make separated Pull Requests, but I got carried out and pushed too much.
- **more tests**
  This PR adds some tests (and travis-ci support!) to yamazaki libraries.
- **add `filename` to Torrent class**
  This modification is needed for some modifications in Miyuki, but it may be useful for several clients who just want to know where are the downloaded torrents.
- **fast `Database#include?` method**
  Track databases will only grow larger and larger, even though the rate is low. It should be noted that clients will perform more searches than insertions, so a faster `#include` is key to a faster Yamazaki. [Yes, 56x faster.](https://gist.github.com/alfateam123/0153df48f5ac1dba5489)
  Unfortunately, the modification requires `Array#bsearch`, which is available only on ruby 2.x. Let me know if you think it has to be compatible with 1.9.x rubies (still shipped by LTS linux distros), we can get around it.

winterthediplomat committed:
add some pointers on how to check available commands

winterthediplomat committed:
add `filename` to Torrent, tests, Database#size

Torrent.filename: it'll contain the path a torrent was downloaded to.

Add tests: Yamazaki now has more tests, covering the usual behaviour
of the library. A travis.yml has been added too.

`Database#size` returns the number of saved items inside the database.

winterthediplomat committed:
Revert the version number change

winterthediplomat committed:
add test about negative index and skip test about unique-filename-in-database

also refactor spec_helper a bit

winterthediplomat committed:
Fasten Database#include?

Switching to Array#bsearch (with database sorting)
makes yamazaki a lot faster (25-56x)

winterthediplomat committed:
Remove ruby 1.9.3 from travis, fix Roxas suggestions

ruby1.9.3 is removed because `Array#bsearch`
is not supported from <2.x ruby releases
Roxas noticed a lot of imperfections and little problems,
they're fixed now.

RoxasShadow commented:
I suggest so squash something if you can and eventually fix `ruby1.9.3 does not have Array#bsearch` that's not actually written in the git-way.

RoxasShadow commented on .travis.yml:
```diff
@@ -0,0 +1,5 @@
+nguage: ruby
```
is that a typo?

alfateam123 replied:
yes.

RoxasShadow commented on .travis.yml:
```diff
@@ -0,0 +1,5 @@
+nguage: ruby
+rvm:
+  - 2.0.0
+  - 2.1.2
```
Aren't version like 2.2.0 and 2.2.0 available? Just to be more standard, you know.

alfateam123 replied:
they should be available. will check if something like `- latest` is available.

alfateam123 replied:
Rox, is the next line in this diff ok for you? \>2.2.1

RoxasShadow replied:
Sure it is

RoxasShadow commented on lib/yamazaki/database.rb:
```diff
@@ -20,15 +20,25 @@ def initialize(track_file, save_on_push = true)
-			@db.count { |t| t[:filename] == filename } > 0
+			begin
+				@db.bsearch { |t| t[:filename] >= filename }[:filename] == filename
+			rescue
```
If i recall correctly, Array#bsearch doesn't raise exception. Am I wrong or am I missing something?

alfateam123 replied:
if Array#bsearch can't find an item, `nil` will be returned. `NilClass` does not have the `[]` operator.
a different way to do it could be
```rb
res = @db.bsearch { |t| t[:filename] >= filename }
if res
  res[:filename] == filename
else
  false
end
```

RoxasShadow replied:
Oh, I got it. Well, let's keep that as it is.

RoxasShadow commented on lib/yamazaki/torrent.rb:
```diff
@@ -15,13 +15,16 @@
 			@pub_date    = pub_date
 			@link        = link
 			@index       = index
+                        @filename    = filename
```
Check the alignment here.

RoxasShadow commented on lib/yamazaki/torrent.rb:
```diff
@@ -15,13 +15,16 @@
 	class Torrent
 		attr_reader :title, :description, :pub_date, :link, :index
+		attr_accessor :filename
+		def initialize(title, description, pub_date, link, index = 0, filename="")
```
Please leave a blank line between 18 and 19.

RoxasShadow commented on specs/database_spec.rb:
```diff
@@ -0,0 +1,91 @@
+  describe '#new' do
+    context 'when track file does not exist' do
+      it 'is created correctly' do
+        File.rm "/tmp/yam.db" if File.exists? "/tmp/yam.db"
```
I suggest to use the current path to avoid platform compatibility issues.

alfateam123 replied:
will do. are you really sure someone will use yamazaki on windows?

RoxasShadow replied:
Absolutely nope, but why losing the support for a platform just for a path? :D

RoxasShadow commented on specs/database_spec.rb:
```diff
@@ -0,0 +1,91 @@
+      it 'is loaded correctly if `track_file` is empty' do
+        track_file = prepare_db([])
+        db = Yamazaki::Database.new(track_file)
+        expect(db.size).to eq(0) 
```
`expect(db).to be_empty` should work too.

RoxasShadow commented on specs/spec_helper.rb:
```diff
@@ -1 +1,28 @@
 require 'yamazaki'
+require 'date'
+require 'base64'
+#require 'random'
```
We can remove this require if we don't need it.

RoxasShadow commented on specs/spec_helper.rb:
```diff
@@ -1 +1,28 @@
+
+def prepare_db(torrents)
+  db_file_path = new_dbfile_path()
+  if torrents.size > 0
```
`any?`

alfateam123 commented on specs/database_spec.rb:
```diff
@@ -0,0 +1,91 @@
+
+
+      it 'does not insert the same filename if already present' do
+        skip 'skip: if needed, a file can be re-downloaded, and we want it to be possible'
```
I'd like to discuss this, though. Should we remove this check?

RoxasShadow replied:
I think filenames should be unique. Why we would suppose the contrary?

RoxasShadow commented:
For me now we're ok :+1:  @hydride0?

alfateam123 commented:
Please wait until a patch for [this issue](https://github.com/hydride0/yamazaki/pull/14#discussion_r42310366) is added (test has to be un-skipped and code modified to guarantee uniqueness)

winterthediplomat committed:
Fix database#<<, now filenames are unique

Filenames should be considered unique in the database,
so duplicates are not allowed. If duplicates are present in
the source file, they won't be deleted after the JSON load
(and it would be very, very time consuming to do this, unless
we switch to hashtables).

winterthediplomat committed:
Improve fake database creation and removal

Fake database files were created under `/tmp`, and
it may break compatibility with non-posix platform.
Moreover they were not deleted, leaving a lot of random
files around: now they are.

alfateam123 commented:
Ok, fixed the filename non-uniqueness. Ok for me too.

RoxasShadow commented:
Good, can you also squash something too? There a lots of commits there. Then let's wait for @hydride0.

alfateam123 commented:
squashed and forced things, is it ok for you @RoxasShadow ?

RoxasShadow commented:
:+1:

alfateam123 commented:
\o/
